### PR TITLE
Documentation and playbook for writable snapshot

### PR DESCRIPTION
**Pull request by Kritika-Bhateja-03**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, pep8, linting, or security issues
- [ ] I have performed Ansible Sanity test using --docker default
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: plugins/modules/writable_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    description:
+    - List of writable snapshot details.
+    required: False
+    type: list
+    elements: dict
+    suboptions:
+      state:
+        description:
+        - The state of the writable snapshot to create or delete.
+        - I(state) is C(present) - To create a writable snapshot.
```
Comment (Bhavneet-Sharma) [inline plugins/modules/writable_snapshot.py:40]: what about get operation?

File: plugins/modules/writable_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    api_password: "{{ api_password }}"
+    writable_snapshot:
+      - dst_path: "/ifs/test_one"
+        src_snap: 2
+        state: present
+      - dst_path: "/ifs/test_two"
+        src_snap: 2
+        state: present
+
+- name: To create a writable snapshot using Name.
```
Comment (Bhavneet-Sharma) [inline plugins/modules/writable_snapshot.py:74]:
```suggestion
- name: To create a writable snapshot using Name
```

File: plugins/modules/writable_snapshot.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    verify_ssl: "{{ verify_ssl }}"
+    api_user: "{{ api_user }}"
+    api_password: "{{ api_password }}"
+    writable_snapshot:
+      - dst_path: "/ifs/test_one"
+        state: absent
+      - dst_path: "/ifs/test_two"
+        sstate: absent
+
+- name: To create and delete writable snapshot
```
Comment (Bhavneet-Sharma) [inline plugins/modules/writable_snapshot.py:100]:
```suggestion
- name: To create and delete writable snapshot
```

File: plugins/modules/writable_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        src_path:
+            description: The directory path of the source snapshot.
+            type: str
+            sample: "/ifs/tfacc_file_system_test"
+        src_snap:
+            description: The directory path of the source snapshot.
+            type: str
+            sample: "Snapshot: 2024Apr15, 4:40 PM"
+        state:
+            description: The name of the source snapshot.s
```
Comment (Bhavneet-Sharma) [inline plugins/modules/writable_snapshot.py:159]:
```suggestion
            description: The name of the source snapshot.
```

File: plugins/modules/writable_snapshot.py (hunk 1 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshot.py:40]: As per spec , I see only create and delete are supported.
Also, get operation would be a part of info module.

File: plugins/modules/writable_snapshot.py (hunk 2 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshot.py]: done

File: plugins/modules/writable_snapshot.py (hunk 4 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshot.py]: done

File: plugins/modules/writable_snapshot.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        """
+
+        ansible_module_params = {
+            'argument_spec': self.get_writable_snapshot_parameters(),
+            'supports_check_mode': True
+        }
+        super().__init__(AnsibleModule, ansible_module_params)
+
+        self.result.update({
+            "writable_snapshot_details": {}
```
Comment (sachin-apa) [inline plugins/modules/writable_snapshot.py:205]:
```suggestion
            "writable_snapshots_details": {}
```

Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshot.py]: done

File: plugins/modules/writable_snapshots.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        try:
+            snapshot_out = self.snapshot_api.get_snapshot_writable_wspath(
+                snapshot_writable_wspath=dst_path
+            ).to_dict().get("writable")
+            writable_snapshot_exists = True if snapshot_out else False
+            return writable_snapshot_exists, snapshot_out[0]
+        except Exception:
+            return False, []
+
+    def create_filesystem_snapshot(self, snapshots_to_create):
```
[the remark this answers is not in the document]
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshots.py]: done

**Review by felixs88 — APPROVED** [review]

**Review by trisha-dell — DISMISSED** [review]

**Comment by Kritika-Bhateja-03** [comment]
UT coverage
image

**Review by felixs88 — APPROVED** [review]

**Review by trisha-dell — DISMISSED** [review]

File: plugins/modules/writable_snapshots.py [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+  - dellemc.powerscale.powerscale
+
+author:
+- Kritika Bhateja(@Kritika-Bhateja-03) <ansible.team.dell.com>
+
+options:
+  writable_snapshots:
+    description:
+    - List of writable snapshots details.
+    required: False
```
Comment (sachin-apa) [inline plugins/modules/writable_snapshots.py:33]:
```suggestion
    required: false
```

File: plugins/modules/writable_snapshots.py [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+        required: true
+        type: path
+      src_snap:
+        description:
+        - The source snapshot name or ID.
+        - This option is required I(state) is C(present).
+        required: false
+        type: str
+notes:
+- The I(check_mode) is supported.
```
Comment (sachin-apa) [inline plugins/modules/writable_snapshots.py:58]: check_mode and diff should be moved to attributes and remove from notes as per the ansible documentation.
https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block 

@anupamaloke FYI.

File: plugins/modules/writable_snapshots.py [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+      - dst_path: "/ifs/test_one"
+        state: absent
+'''
+
+RETURN = r'''
+changed:
+    description: Whether or not the resource has changed.
+    returned: always
+    type: bool
+    sample: "true"
```
Comment (sachin-apa) [inline plugins/modules/writable_snapshots.py:123]:
```suggestion
    sample: true
```

File: plugins/modules/writable_snapshots.py (hunk 7 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshots.py]: done

File: plugins/modules/writable_snapshots.py (hunk 8 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshots.py]: done

File: plugins/modules/writable_snapshots.py (hunk 9 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/writable_snapshots.py]: done

**Review by felixs88 — DISMISSED** [review]

**Review by sachin-apa — DISMISSED** [review]

File: tests/unit/plugins/module_utils/mock_writable_snapshots_api.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+# Copyright: (c) 2022-2024, Dell Technologies
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/module_utils/mock_writable_snapshots_api.py:1]:
```suggestion
# Copyright: (c) 2024, Dell Technologies
```

File: tests/unit/plugins/module_utils/mock_writable_snapshots_api.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""Mock Api response for Unit tests of writablesnapshots module on PowerScale"""
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/module_utils/mock_writable_snapshots_api.py:5]:
```suggestion
"""Mock Api response for Unit tests of writable snapshots module on PowerScale"""
```

File: plugins/modules/writable_snapshots.py [hunk 12]
```diff
@@ -0,0 +1,596 @@
...
+        - The state of the writable snapshot to create or delete.
+        - I(state) is C(present) - To create a writable snapshot.
+        - I(state) is C(absent) - To delete the writable snapshot.
+        default: present
+        choices: ['present', 'absent']
+        type: str
+      dst_path:
+        description:
+        - The /ifs of the writable snapshot.
+        - The destination path should be non-existing path and it's absolute path.
```
Comment (Bhavneet-Sharma) [inline plugins/modules/writable_snapshots.py:48]:
```suggestion
        - The destination path should be non-existing path and it is absolute path.
```

**Review by sachin-apa — APPROVED** [review]

File: docs/modules/writable_snapshots.rst [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+      \ :emphasis:`state`\  is \ :literal:`present`\  - To create a writable snapshot.
+
+      \ :emphasis:`state`\  is \ :literal:`absent`\  - To delete the writable snapshot.
+
+
+    dst_path (True, path, None)
+      The /ifs of the writable snapshot.
+
+      The destination path should be non-existing path and it's absolute path.
```
Comment (Bhavneet-Sharma) [inline docs/modules/writable_snapshots.rst:53]:
```suggestion
      The destination path should be non-existing path and it is absolute path.
```

**Review by Bhavneet-Sharma — APPROVED** [review]

**Review by trisha-dell — APPROVED** [review]

**Pull request merged**